### PR TITLE
Fix MariaDB list columns performance

### PR DESCRIPTION
**Pull request by ausi**
|      Q       |   A
|------------- | -----------
| Type         | bug
| Fixed issues | https://github.com/contao/contao/issues/6409

#### Summary

Starting with version 3.7.0 we noticed that the `listTableColumns()` method got very slow on some systems, our migrate command went from taking around 2 seconds with version 3.6 to 260 seconds with version3.7. 

This is due to the `LEFT JOIN information_schema.CHECK_CONSTRAINTS` not getting optimized by MariaDB which causes the database server to scan every table of every database.

This pull request fixes that by using a subquery that is limited by `WHERE CONSTRAINT_SCHEMA = $databaseName`. This brings the performance back to where it was with version 3.6.

### CODE REVIEW

File: src/Platforms/MariaDb1043Platform.php [hunk 1]
```diff
@@ -76,30 +76,33 @@ public function getListTableColumnsSQL($table, $database = null): string
...
 
+        $databaseName = $this->getDatabaseNameSQL($databaseName);
+
         $columnTypeSQL = <<<SQL
             IF(
-                x.CHECK_CLAUSE IS NOT NULL AND $tableAlias.COLUMN_TYPE = 'longtext',
+                $tableAlias.COLUMN_TYPE = 'longtext'
+                AND EXISTS(
+                    SELECT * from information_schema.CHECK_CONSTRAINTS 
+                    WHERE CONSTRAINT_SCHEMA = $databaseName
```
Comment (ausi) [inline src/Platforms/MariaDb1043Platform.php:92]: I tried using `WHERE CONSTRAINT_SCHEMA = $tableAlias.TABLE_SCHEMA` here to simplify the PHP code, but sadly this seems to prevent the optimization in MariaDB.

**Review by fritzmg — APPROVED** [review]

**Comment by fritzmg** [comment]
Just to add more notes: this is especially noticeable in local development environments where you may have a lot of databases with lots of tables from various projects - as the unoptimized `information_schema.CHECK_CONSTRAINTS` query will then scan _every_ table of _every_ database _every_ time, which is very slow. Thus exacerbating the issue when `listTableColumns()` is executed multiple times during database migrations as previously mentioned (hence the operation time sky rocketing from just 2 seconds to 260 seconds for example).

**Review by m-vo — APPROVED** [review]

File: src/Platforms/AbstractMySQLPlatform.php [hunk 2]
```diff
@@ -404,7 +404,7 @@ public function getListTableColumnsSQL($table, $database = null)
      *
      * @return array{string, string}
      */
-    public function getColumnTypeSQLSnippets(string $tableAlias = 'c'): array
+    public function getColumnTypeSQLSnippets(string $tableAlias = 'c', ?string $databaseName = null): array
```
Comment (derrabus) [inline src/Platforms/AbstractMySQLPlatform.php]: Our platform classes are meant to be extended, so adding new parameters to methods is a BC break unfortunately. Can we solve this issue without introducing new parameters?

File: tests/Functional/Schema/MySQLSchemaManagerTest.php [hunk 3]
```diff
@@ -401,6 +402,11 @@ public function testJsonColumnType(): void
         $table->addColumn('col_json', Types::JSON);
         $this->dropAndCreateTable($table);
 
+        // Remove the comment from the column to ensure the type is detected correctly from the check constraints.
+        if ($this->connection->getDatabasePlatform() instanceof MariaDb1043Platform) {
+            $this->connection->executeStatement('ALTER TABLE test_mysql_json CHANGE COLUMN col_json col_json JSON');
+        }
```
Comment (derrabus) [inline tests/Functional/Schema/MySQLSchemaManagerTest.php:408]: Why is this change necessary?

**Review by derrabus — COMMENTED** [review]
Sorry for picking this PR up so late.

File: tests/Functional/Schema/MySQLSchemaManagerTest.php (hunk 3 above)
Comment (ausi) [reply tests/Functional/Schema/MySQLSchemaManagerTest.php:408]: This is needed to make sure the code in `getColumnTypeSQLSnippets()` gets actually tested. Before it was untested but I wanted to make sure that my modifications don’t break it.

It removes the `(DC2Type:json)` comment from the column, making sure the JSON type is still correctly detected by looking up the `information_schema.CHECK_CONSTRAINTS`.

File: src/Platforms/AbstractMySQLPlatform.php (hunk 2 above)
Comment (ausi) [reply src/Platforms/AbstractMySQLPlatform.php]: I tried, but I was not able to find a solution without the added parameter. But as this method ist very new (first released with 3.7.0) I’d guess that the BC impact would be very low.

If you decide against adding the parameter we could use `func_get_arg(1)` and document the new parameter, or create a new method that just calls this one by default.

**Comment by ausi** [comment]
> Sorry for picking this PR up so late.

Nothing to be sorry about! Thank you for all your work on doctrine!

File: src/Platforms/AbstractMySQLPlatform.php (hunk 2 above)
Comment (derrabus) [reply src/Platforms/AbstractMySQLPlatform.php]: After your change, all implementations of `getColumnTypeSQLSnippets()` will return a tuple where the second element is an empty string. So we're effectively not returning snippets (plural) anymore. So we're better off creating a new method with a proper name.

Comment (ausi) [reply src/Platforms/AbstractMySQLPlatform.php]: But that would be a BC break again, as we cannot redirect the new method to the old method then. All overwritten `getColumnTypeSQLSnippets()` methods wouldn’t get used anymore.

But I’m happy either way, which one should I implement?

Comment (derrabus) [reply src/Platforms/AbstractMySQLPlatform.php]: If we can't fix this issue without breaking BC, then so be it. 😕 

Add a new method and document the BC break that overriding `getColumnTypeSQLSnippets()` won't have any effect anymore. And please target 3.8.x for that change. It's to heavy for a patch release then.

Comment (ausi) [reply src/Platforms/AbstractMySQLPlatform.php]: That would make the 3.7 version unusable for all our users affected by this.
In this case I’d favor using `func_get_arg(1)` (AFAIK Symfony uses this approach often for BC reasons), that would be 100% backwards compatible.

Comment (ausi) [reply src/Platforms/AbstractMySQLPlatform.php]: Implemented in d0ac1daf7c48b424a92d88e5f4a41ea22901e210
@derrabus what do you think?

Comment (derrabus) [reply src/Platforms/AbstractMySQLPlatform.php]: The issue that `getColumnTypeSQLSnippets()` is a broken API now remains, however:

* Calling the method without arguments is not desirable.
* The method effectively returns only one snippet, so its name and return type don't fit anymore.

How about adding a new method with the API we want and make the current method call the new one? On 3.8.x, we can deprecate the `getColumnTypeSQLSnippets()` method in favor of your new one.

Comment (ausi) [reply src/Platforms/AbstractMySQLPlatform.php]: Done in 82d8266551ffa10dad6b5542eff980dc9002b867

I added the deprecation for `getColumnTypeSQLSnippets()` directly as I don’t think we need to wait for this until version 3.8, do we?

Comment (derrabus) [reply src/Platforms/AbstractMySQLPlatform.php]: > I don’t think we need to wait for this until version 3.8, do we?

If you want this to be merged to 3.7.x, we do. No new deprecations in bugfix releases. 😇 

But I don't mind taking this directly to 3.8, if that makes it easier for you.

Comment (ausi) [reply src/Platforms/AbstractMySQLPlatform.php]: No, we need it in 3.7 ☺️
I will revert the deprecation and create a new pr for 3.8

**Comment by ausi** [comment]
This should be ready now.

Followup pull requests:
1. Deprecation in 3.8: https://github.com/doctrine/dbal/pull/6213
2. Removal in 4.0: https://github.com/doctrine/dbal/pull/6214

**Review by derrabus — APPROVED** [review]
